lib/virtual.js

```diff
@@ -50,39 +122,100 @@ function VirtualModulesPlugin(compiler) {
 }
 
 VirtualModulesPlugin.prototype.writeModule = function(filePath, contents) {
-	var len = contents ? contents.length : 0;
-	var time = new Date();
-	var timeMs = time.getTime();
+	const len = contents ? contents.length : 0;
+	const stats = createVirtualStats(len);
+	const modulePath = getModulePath(filePath, this.compiler);
 
-	var stats = new VirtualStats({
-		dev: 8675309,
-		ino: inode++,
-		mode: 33188,
-		nlink: 0,
-		uid: 1000,
-		gid: 1000,
-		rdev: 0,
-		size: len,
-		blksize: 4096,
-		blocks: Math.floor(len / 4096),
-		atimeMs: timeMs,
-		mtimeMs: timeMs,
-		ctimeMs: timeMs,
-		birthtimeMs: timeMs,
-		atime: time,
-		mtime: time,
-		ctime: time,
-		birthtime: time
-	});
+	// When using the WatchIgnorePlugin (https://github.com/webpack/webpack/blob/52184b897f40c75560b3630e43ca642fcac7e2cf/lib/WatchIgnorePlugin.js),
+	// the original watchFileSystem is stored in `wfs`. The following "unwraps" the ignoring
+	// wrappers, giving us access to the "real" watchFileSystem.
+	let finalWatchFileSystem = this._watcher && this._watcher.watchFileSystem;
 
+	while (finalWatchFileSystem && finalWatchFileSystem.wfs) {
+		finalWatchFileSystem = finalWatchFileSystem.wfs;
+	}
 	this.compiler.inputFileSystem._writeVirtualFile(filePath, stats, contents);
+	if (finalWatchFileSystem &&
+		(finalWatchFileSystem.watcher.fileWatchers.size ||
+			finalWatchFileSystem.watcher.fileWatchers.length)
+	) {
+		const fileWatchers = finalWatchFileSystem.watcher.fileWatchers instanceof Map ?
+			Array.from(finalWatchFileSystem.watcher.fileWatchers.values()) :
+			finalWatchFileSystem.watcher.fileWatchers;
+		fileWatchers.forEach(function(fileWatcher) {
+			if (fileWatcher.path === modulePath) {
+				delete fileWatcher.directoryWatcher._cachedTimeInfoEntries;
+				fileWatcher.directoryWatcher.setFileTime(
+					filePath,
+					stats.birthtime,
+					false,
+					false,
+					null
+				);
+				fileWatcher.emit("change", stats.birthtime, null);
+			}
+		});
+	}
 };
 
-function setData(storage, key, value) {
-	if (storage.data instanceof Map) {
-		storage.data.set(key, value);
+function getData(storage, key) {
+	const data = storage._data /* webpack 5 */ || storage.data /* webpack 4 */;
+	if (data instanceof Map) {
+		return storage.data.get(key);
+	}
+	return data.data[key];
+}
+
+function setData(backendOrStorage, key, valueFactory) {
```

Comment:
This implementation is not nice. Better looks like getData, a webpack switch and a generic code